core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,449 @@
+    Either.catchNonFatal(unsafeMapIntCodePoint(codePoint)).leftMap(_.getLocalizedMessage)
+
+  /**
+   * An error that is yielded when mapping an individual code point fails.
+   */
+  sealed abstract class CodePointMappingException extends RuntimeException with NoStackTrace {
+
+    /**
+     * The index of the start of the Unicode code point in the input where the failure occurred.
+     */
+    def failureIndex: Int
+
+    /**
+     * A description of why the failure occurred.
+     */
+    def message: String
+
+    /**
+     * The [[CodePoint]] which caused the failure.
+     */
+    def codePoint: CodePoint
+
+    // final
+
+    final override def getMessage: String =
+      toString
+
+    final override def toString: String =
+      s"CodePointMappingException(message = $message, failureIndex = $failureIndex, codePoint = $codePoint)"
+  }
+
+  object CodePointMappingException {
+    final private[this] case class CodePointMappingExceptionImpl(
+        override val failureIndex: Int,
+        override val message: String,
+        override val codePoint: CodePoint)
+        extends CodePointMappingException
+
+    def apply(
+        failureIndex: Int,
+        message: String,
+        codePoint: CodePoint): CodePointMappingException =
+      CodePointMappingExceptionImpl(failureIndex, message, codePoint)
+
+    implicit val hashAndOrderForCodePointMappingException
+        : Hash[CodePointMappingException] with Order[CodePointMappingException] =
+      new Hash[CodePointMappingException] with Order[CodePointMappingException] {
+        override def hash(x: CodePointMappingException): Int = x.hashCode
+
+        override def compare(x: CodePointMappingException, y: CodePointMappingException): Int =
+          x.failureIndex.compare(y.failureIndex) match {
+            case 0 =>
+              x.message.compare(y.message) match {
+                case 0 =>
+                  x.codePoint.compare(y.codePoint)
+                case otherwise => otherwise
+              }
+            case otherwise => otherwise
+          }
+      }
+
+    implicit val showForCodePointMappingException: Show[CodePointMappingException] =
+      Show.fromToString
+  }
+
+  /**
+   * An error which is yielded if attempting to map a sequence of Unicode code points with the
+   * UTS-46 mapping algorithm fails.
+   */
+  sealed abstract class MappingException extends RuntimeException with NoStackTrace {
+
+    /**
+     * One or more mapping [[CodePointMappingException]].
+     *
+     * Each [[CodePointMappingException]] describes the mapping failure of an independent code
+     * point.
+     */
+    def errors: NonEmptyList[CodePointMappingException]
+
+    /**
+     * The input string, mapped as much as was possible. Code points which failed replaced with
+     * the Unicode replacement character � (0xFFFD). Returning this value on failure is mandated
+     * by UTS-46.
+     */
+    def partiallyMappedInput: String
```

Comment:
Interesting that this is mandated. I pondered whether our return type should be shaped more like `Ior` or `(Option[MappingException], String)` but I think the `Either` is best understood.

Reply:
I don't want to use `Ior` because the `Ior.both` is a bit ambiguous on if the right side is really success or not. In this case, the `partiallyMappedInput` is definitely a failure.